server/src/main/java/org/elasticsearch/index/translog/Translog.java

```diff
@@ -1093,47 +1094,50 @@ public static Type fromId(byte id) {
             }
         }
 
-        Type opType();
+        protected final long seqNo;
 
-        long estimateSize();
+        protected final long primaryTerm;
 
-        BytesReference source();
+        protected Operation(long seqNo, long primaryTerm) {
+            this.seqNo = seqNo;
+            this.primaryTerm = primaryTerm;
+        }
+
+        public abstract Type opType();
 
-        long seqNo();
+        public abstract long estimateSize();
+
+        public final long seqNo() {
+            return seqNo;
+        }
 
-        long primaryTerm();
+        public final long primaryTerm() {
+            return primaryTerm;
+        }
 
         /**
-         * Reads the type and the operation from the given stream. The operation must be written with
-         * {@link Operation#writeOperation(StreamOutput, Operation)}
+         * Reads the type and the operation from the given stream.
          */
-        static Operation readOperation(final StreamInput input) throws IOException {
+        public static Operation readOperation(final StreamInput input) throws IOException {
             final Translog.Operation.Type type = Translog.Operation.Type.fromId(input.readByte());
             return switch (type) {
                 // the de-serialization logic in Index was identical to that of Create when create was deprecated
-                case CREATE, INDEX -> new Index(input);
-                case DELETE -> new Delete(input);
+                case CREATE, INDEX -> Index.readFrom(input);
```

Comment:
I think defining an abstract `readFrom()` method similar to what you did with `writeBody()` will simplify this a lot and remove the need for the `switch` here (which is weird, since the abstract class would need to know all possible sub classes). What do you think?

Reply:
But we still have the switch in that abstract method then don't we? 
I couldn't really find a nicer way than this (just moving the method seems like redundant noise?) other than maybe putting the read method on the type enum to avoid manual dispatch, but in practice that was more code.

I'll leave it as is for now but maybe we can make it nicer in a follow-up if I missed something :)

Reply:
> which is weird, since the abstract class would need to know all possible sub classes 

We can enforce this at compile time these days, see https://github.com/elastic/elasticsearch/pull/95109